libs/datasets/data_source.py

```diff
@@ -1,59 +1,65 @@
-from typing import Type
+import pathlib
+from typing import List
+from typing import Optional
+from typing import Union
 
-import pandas as pd
+import structlog
+from covidactnow.datapublic import common_df
+from covidactnow.datapublic.common_fields import CommonFields
 
-from libs.datasets.dataset_utils import STATIC_INDEX_FIELDS
+from libs.datasets import dataset_utils
+from libs.datasets import timeseries
 from libs.datasets.dataset_utils import TIMESERIES_INDEX_FIELDS
 from libs.datasets.timeseries import MultiRegionDataset
 from functools import lru_cache
+import pandas as pd
 
+_log = structlog.get_logger()
 
-class DataSource(object):
-    """Represents a single dataset source, loads data and cleans data."""
-
-    # Subclass must implement custom class of fields in dataset.
-    class Fields(object):
-        pass
-
-    INDEX_FIELD_MAP = None
 
-    COMMON_FIELD_MAP = None
+class DataSource(object):
+    """Represents a single dataset source; loads data and produces a MultiRegionDataset."""
 
-    # Name of dataset source
+    # Name of source
+    # TODO(tom): Make an enum of these.
     SOURCE_NAME = None
 
-    # Indicates if NYC data is aggregated into one NYC county or not.
-    HAS_AGGREGATED_NYC_BOROUGH = False
-
-    def __init__(self, data: pd.DataFrame):
-        self.data = data
+    # Fields expected to be in the DataFrame loaded by common_df.read_csv
+    EXPECTED_FIELDS: Optional[List[CommonFields]] = None
 
-    @classmethod
-    @lru_cache(None)
-    def local(cls) -> "DataSource":
-        """Builds data from local covid-public-data github repo.
+    # Path of the CSV to be loaded by the default `make_dataset` implementation.
+    COMMON_DF_CSV_PATH: Optional[Union[pathlib.Path, str]] = None
 
-        Returns: Instantiated class with data loaded.
-        """
-        raise NotImplementedError("Subclass must implement")
+    # Fields that are ignored when warning about missing and extra fields. By default some fields
+    # that contain redundant information about the location are ignored because cleaning them up
+    # isn't worth the effort.
+    IGNORED_FIELDS = (CommonFields.COUNTY, CommonFields.COUNTRY, CommonFields.STATE)
 
+    @classmethod
     @lru_cache(None)
-    def multi_region_dataset(self) -> MultiRegionDataset:
-        if set(self.INDEX_FIELD_MAP.keys()) == set(TIMESERIES_INDEX_FIELDS):
-            dataset = MultiRegionDataset.from_fips_timeseries_df(self.data).add_provenance_all(
-                self.SOURCE_NAME
+    def make_dataset(cls) -> timeseries.MultiRegionDataset:
+        """Default implementation of make_dataset that loads timeseries data from a CSV."""
+        assert cls.COMMON_DF_CSV_PATH, f"No path in {cls}"
+        data_root = dataset_utils.LOCAL_PUBLIC_DATA_PATH
+        input_path = data_root / cls.COMMON_DF_CSV_PATH
+        data = common_df.read_csv(input_path, set_index=False)
+        expected_fields = pd.Index({*cls.EXPECTED_FIELDS, *TIMESERIES_INDEX_FIELDS})
```

Comment:
It would be great to trim down the `TIMESERIES_INDEX_FIELDS` to only include fips and date - we don't need most of the fields and we can get aggregate level from the region object itself

Reply:
Created https://trello.com/c/2Pa4DMyu/836-trim-down-the-timeseriesindexfields